examples/train_model.py

```diff
@@ -128,12 +128,20 @@ def main(parser):
     impatience = 0
     saved = False
     valid_world = None
+    num_epochs = 0
+    random = opt.get('datatype', 'train') == 'train'
     with world:
         while True:
             world.parley()
             parleys += 1
 
-            if opt['num_epochs'] > 0 and parleys >= max_parleys:
+            if world.epoch_done():
+                num_epochs += 1
+            elif type(world) is BatchWorld:
```

Comment:
we shouldnt need a batchworld check inside this loop, how can we avoid this?
this looks pretty horrible and points to something being wrong in our design if required :(

Reply:
this script should be a simple entry point into parlai, not ridiculously complicated..

Reply:
I'll move the logic out of here into BatchWorld - sorry about that